src/manifest.rs

```diff
@@ -33,15 +33,212 @@ use crate::{
 /// The name of the manifest file
 pub const MANIFEST_FILE: &str = "Proto.toml";
 
-/// A `buffrs` manifest format used for serialization and deserialization.
+/// The canary edition supported by this version of buffrs
+pub const CANARY_EDITION: &str = concat!("0.", env!("CARGO_PKG_VERSION_MINOR"));
+
+/// Edition of the buffrs manifest
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Edition {
+    /// The canary edition of manifests
+    ///
+    /// This indicates that breaking changes and unstable behavior can occur
+    /// at any time. Users are responsible for consulting documentation and
+    /// help channels if errors occur.
+    Canary,
+    /// Unknown edition of manifests
+    ///
+    /// This is unrecommended as breaking changes could be introduced due to being
+    /// in the beta release channel
+    Unknown,
+}
+
+impl Default for Edition {
+    fn default() -> Self {
+        Self::Canary
+    }
+}
+
+/// A buffrs manifest format used for serialization and deserialization.
 ///
 /// This contains the exact structure of the `Proto.toml` and skips
 /// empty fields.
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
-struct RawManifest {
-    package: Option<PackageManifest>,
-    #[serde(default)]
-    dependencies: DependencyMap,
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum RawManifest {
+    Canary {
+        package: Option<PackageManifest>,
+        dependencies: DependencyMap,
+    },
+    Unknown {
```

Comment:
Genuine question: is `Unknown` useful if you are raising an error when parsing a manifest without edition?

Reply:
This is worth a revisit!

Reply:
Re: Yes.

This is used to omit a re-serialization of the manifest that users use (to make the edition management a virtual thing)

Example:

The manifest that a user hast may be:

```
[package]
type = "lib"
name = "xyz"
version = "1.0.0"
```

which _could_ be (or in fact is) a valid manifest across (or since) many different editions as these keys are enough to publish a lib without any deps.

If we would not deserialize this manifest as "unknown" but as "canary" / "0.7" – we would set the edition after every buffrs command that edits the manifest.

This would resolve in an actual incompatibility (e.g. if 0.7 is incompatible with 0.8 due to a field being added in the dependencies section) even though it is actually compatible / deserializable with the 0.8 schema.

E.g.

```
edition = "0.7"

[package]
type = "lib"
name = "xyz"
version = "1.0.0"
```

would yield an error in case buffrs 0.8 is used in this  project and the user would have to manually update the manifest to be 0.8 here, but there is no migration to be done:

```
edition = "0.8"

[package]
type = "lib"
name = "xyz"
version = "1.0.0"
```

instead when using:

```
[package]
type = "lib"
name = "xyz"
version = "1.0.0"
```

the edition is inferred / treated to be the latest (but internally stored as the `Unknown` varianted and _only_ serialized once a package is being created for the publishing.

This results in the behavior of automatic inlining once `buffrs publish` is called:

(content of `Proto.toml`):

```
[package]
type = "lib"
name = "xyz"
version = "1.0.0"
```

(content of `Proto.toml` in the `lib-1.0.0.tgz`):

```
edition = "0.7"

[package]
type = "lib"
name = "xyz"
version = "1.0.0"
```

Reply:
I hope this was somewhat clear, lmk if I explained it wrong (or in an messy way)